### PR TITLE
Kernel mode support for fnmp/fnlwf

### PR DESCRIPTION
**Pull request by nigriMSFT**
Exposes kernel mode support for fnmpapi.h and fnlwfapi.h. Tests were updated to exercise the kernel mode API.

Implementation updates needed in the fnmp/fnlwf code to get the kernel mode API working:
- Plumb IRP RequestorMode to avoid ProbeForRead calls for kernel buffers

User observable updates:
- Publish Invoke System Relay (Isr) as another test tool
- Remove fnmpapi.dll/fnlwfapi.dll in favor of a header-only API
- Add xplat abstraction for fnmpapi/fnlwfapi handles and return statuses
- Simplify pkthlp.h to avoid re-defining structures already available in Windows headers

Other notable internal changes:
- Refactor common IOCTL handling code into fnioctl_[um|km].h
- Rename fnmp/fnlwf IOCTL symbols to avoid name collision
- Move fnmp configuration defaults to separate header fnmpapiconfig.h
- [test] Add cxplat lib
- [test] Replace C++ STL usage w/ cxplat lib and hand-rolled helpers
- [test] Add fnfunctionaltestdrv driver to execute tests in kernel-mode
- [test] Refactor test logic to gracefully exit upon failure

Addresses #38

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -108,7 +109,7 @@ jobs:
       uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32
       if: ${{ always() }}
       with:
-        name: logs_func_win${{ matrix.windows }}_${{ matrix.configuration }}_${{ matrix.platform }}
+        name: logs_func_${{ matrix.apimode }}_win${{ matrix.os }}_${{ matrix.configuration }}_${{ matrix.platform }}
```
Comment (nibanks) [inline .github/workflows/ci.yml:112]: Just a nit: This whole repo is **win**-net-test. I feel like the Windows part is then implied for everything else. So it's unnecessary to include it in any output.
```suggestion
        name: logs_func_${{ matrix.apimode }}_${{ matrix.os }}_${{ matrix.configuration }}_${{ matrix.platform }}
```

File: src/isr/svc/isrsvc.vcxproj [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  <ItemGroup>
+    <ClCompile Include="main.cpp" />
+  </ItemGroup>
+  <PropertyGroup Condition="'$(DisableRegistryUse)' == 'true'">
+    <!--
+      Workaround for EWDK build failures:
+      Microsoft.Cpp.VCTools.props sets these to invalid values if they're not
+      already set. -->
+    <VCInstallDir_160 Condition="'$(VCInstallDir_160)' == ''">$(VSInstallDir)VC\</VCInstallDir_160>
+    <VCToolsInstallDir_160 Condition="'$(VCToolsInstallDir_160)' == '' and '$(VCToolsVersion)' != ''">$(VCInstallDir_160)Tools\MSVC\$(VCToolsVersion)\</VCToolsInstallDir_160>
```
Comment (nibanks) [inline src/isr/svc/isrsvc.vcxproj:22]: Should these kinds of things be pushed to the undocked repo/props file?

File: src/isr/svc/isrsvc.vcxproj [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      <PreprocessorDefinitions>
+        %(PreprocessorDefinitions)
+      </PreprocessorDefinitions>
+      <UseFullPaths>true</UseFullPaths>
+      <!-- Disable C26812: The enum type '' is unscoped. Prefer 'enum class' over 'enum' -->
+      <!-- Disable C5252 : Multiple different types resulted in the same XFG type-hash -->
+      <DisableSpecificWarnings>26812;5252;%(DisableSpecificWarnings)</DisableSpecificWarnings>
+    </ClCompile>
+    <Link>
+      <SubSystem>Console</SubSystem>
```
Comment (nibanks) [inline src/isr/svc/isrsvc.vcxproj:44]: Should be unnecessary. Can you please try to remove?

File: src/isr/svc/isrsvc.vcxproj [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    </PreBuildEvent>
+    <ClCompile>
+      <AdditionalIncludeDirectories>
+        $(SolutionDir)inc;
+        %(AdditionalIncludeDirectories)
+      </AdditionalIncludeDirectories>
+      <PreprocessorDefinitions>
+        %(PreprocessorDefinitions)
+      </PreprocessorDefinitions>
+      <UseFullPaths>true</UseFullPaths>
```
Comment (nibanks) [inline src/isr/svc/isrsvc.vcxproj:38]: What is this for? Does it need to be here? If so, it this something that should go in the undocking props too?

File: test/cxplat/lib/um/cxplat_um.vcxproj [hunk 5]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <ProjectGuid>{93edb3a0-c6e4-42cc-85a3-92bfc44964d0}</ProjectGuid>
+    <TargetName>cxplat_um</TargetName>
+    <UndockedType>lib</UndockedType>
+    <UndockedDir>$(SolutionDir)submodules\undocked\</UndockedDir>
+    <UndockedOut>$(SolutionDir)artifacts\</UndockedOut>
+    <UndockedSourceLink>true</UndockedSourceLink>
+  </PropertyGroup>
```
Comment (nibanks) [inline test/cxplat/lib/um/cxplat_um.vcxproj:10]: Why don't you set to use the driver toolset? I think that will allow you to use the WPP stuff and not have to run the command yourself. Maybe it helps with the VCInstallDir stuff too?

File: test/functional/bin/km/fnfunctionaltestdrv.vcxproj [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      <WppAdditionalOptions>-p:fntestdrv</WppAdditionalOptions>
+    </ClCompile>
+    <Link>
+      <AdditionalDependencies>
+        netio.lib;
+        uuid.lib;
+        wdmsec.lib;
+        %(AdditionalDependencies)
+      </AdditionalDependencies>
+      <LinkTimeCodeGeneration>UseLinkTimeCodeGeneration</LinkTimeCodeGeneration>
```
Comment (nibanks) [inline test/functional/bin/km/fnfunctionaltestdrv.vcxproj:53]: Necessary? If so, can/should it move to the undocked props?

File: test/functional/bin/km/fnfunctionaltestdrv.vcxproj [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  <ItemGroup>
+    <ClCompile Include="control.cpp" />
+    <ClCompile Include="driver.cpp" />
+    <ResourceCompile Include="fnfunctionaltestdrv.rc" />
+  </ItemGroup>
+  <PropertyGroup Label="Configuration">
+    <TargetVersion>Windows10</TargetVersion>
+    <PlatformToolset>WindowsKernelModeDriver10.0</PlatformToolset>
+    <ConfigurationType>Driver</ConfigurationType>
+    <DriverType>KMDF</DriverType>
```
Comment (nibanks) [inline test/functional/bin/km/fnfunctionaltestdrv.vcxproj:31]: All these already come from the undocked props.

File: test/functional/bin/km/fnfunctionaltestdrv.vcxproj [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    <DriverTargetPlatform>Universal</DriverTargetPlatform>
+    <_NT_TARGET_VERSION>0x0A00000A</_NT_TARGET_VERSION>
```
Comment (nibanks) [inline test/functional/bin/km/fnfunctionaltestdrv.vcxproj:33]: Are these actually necessary?

File: test/functional/lib/fnfunctionaltestlib_um.vcxproj [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        %(AdditionalIncludeDirectories)
+      </AdditionalIncludeDirectories>
+      <PreprocessorDefinitions>
+        UM_NDIS683=1;
+        %(PreprocessorDefinitions)
+      </PreprocessorDefinitions>
+      <UseFullPaths>true</UseFullPaths>
+      <!-- Disable C26812: The enum type '' is unscoped. Prefer 'enum class' over 'enum' -->
+      <!-- Disable C5252 : Multiple different types resulted in the same XFG type-hash -->
+      <DisableSpecificWarnings>26812;5252;%(DisableSpecificWarnings)</DisableSpecificWarnings>
```
Comment (nibanks) [inline test/functional/lib/fnfunctionaltestlib_um.vcxproj:47]: I've seen you do this in several projects. Should it be moved to `wnt.cpp.props`?

**Review by nibanks — APPROVED** [review]

**Review by nibanks — APPROVED** [review]

File: test/cxplat/lib/um/cxplat_um.vcxproj (hunk 5 above)
Comment (nigriMSFT) [reply test/cxplat/lib/um/cxplat_um.vcxproj:10]: I attempted this and ran into some interesting linking issues with runtime libs that I could not resolve in a reasonable amount of time. So I think I will leave this be for now

File: src/isr/svc/isrsvc.vcxproj (hunk 2 above)
Comment (nigriMSFT) [reply src/isr/svc/isrsvc.vcxproj]: This was copied from somewhere or added by someone else, I don't quite understand the purpose of this. I've removed all instances. We will see after this gets checked in if they are needed for the OB build

File: test/functional/lib/fnfunctionaltestlib_um.vcxproj (hunk 9 above)
Comment (nigriMSFT) [reply test/functional/lib/fnfunctionaltestlib_um.vcxproj]: Yep, moved

File: test/cxplat/lib/um/cxplat_um.vcxproj (hunk 5 above)
Comment (nibanks) [reply test/cxplat/lib/um/cxplat_um.vcxproj:10]: Sounds good. Thanks for trying.

**Review by nibanks — APPROVED** [review]

**Pull request merged**